File: modules/13b-alignment/ts/02-reward-model.ts
/**
 * Task 2 🟡 — Reward model from preferences (Bradley–Terry).
 *
 * What you'll learn:
 *   - The Bradley–Terry model: turning "chosen ≻ rejected" pairs into a
 *     probabilistic training signal
 *   - The reward-modeling loss used by InstructGPT: −log σ(r_chosen − r_rejected)
 *   - Why maximum likelihood on noisy preferences recovers the hidden reward
 *     (direction) — the exact recipe behind every RLHF reward model
 *
 * The math (README derives each step):
 *
 *   Responses are feature vectors x ∈ ℝ^D. A hidden true reward scores them:
 *       r*(x) = w* · x
 *   Preference data: for a pair (x_a, x_b), the label "a chosen" is sampled with
 *       P(a ≻ b) = σ(r*(x_a) − r*(x_b))          (Bradley–Terry)
 *   We fit a linear reward model  r_θ(x) = θ · x  by minimising, over pairs,
 *       L(θ) = −mean log σ(Δ) ,   Δ = r_θ(x_chosen) − r_θ(x_rejected)
 *   Gradient (derive it: d/dΔ [−log σ(Δ)] = −(1 − σ(Δ)), then chain to θ):
 *       ∂L/∂θ = −mean (1 − σ(Δ)) · (x_chosen − x_rejected)
 *
 * You implement the three core functions (btProb, btLoss, btGradStep) using
 * plain arrays only (no math libraries). Data generation, the train/held-out
 * split, the training loop, and the evaluation report are provided and
 * runnable.
 *
 * How to run:
 *   pnpm tsx modules/13b-alignment/ts/02-reward-model.ts
 */

const SEED = 21;
const D = 6; // response feature dimension
const N_TRAIN = 400; // preference pairs for training
const N_TEST = 200; // held-out preference pairs
const LR = 0.2;
const EPOCHS = 200;

// The hidden "true" reward direction the annotators (noisily) follow.
const W_TRUE = [3.0, -4.0, 2.0, 1.2, -2.4, 1.6];

// ---------------------------------------------------------------------------
// Seeded RNG (provided) — LCG + Box-Muller for reproducible Gaussians
// ---------------------------------------------------------------------------

function makeRng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    // Numerical Recipes LCG
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296; // uniform in [0, 1)
  };
}

function makeGaussianFrom(u: () => number): () => number {
  return () => {
    // Box-Muller: two uniforms → one standard normal
    const u1 = u() + 1e-12;
    const u2 = u();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  };
}

// ---------------------------------------------------------------------------
// Small vector helpers (provided)
// ---------------------------------------------------------------------------

/** Dot product a·b. */
function dot(a: number[], b: number[]): number {
  return a.reduce((acc, v, i) => acc + v * b[i], 0);
}

/** Euclidean norm ‖a‖. */
function norm(a: number[]): number {
  return Math.sqrt(dot(a, a));
}

// ---------------------------------------------------------------------------
// Preference-pair generation  (provided — do not edit)
// ---------------------------------------------------------------------------

/**
 * Generate n preference pairs. Each pair draws two random "responses"
 * xA, xB ~ N(0, I_D); the winner is sampled from the Bradley–Terry
 * probability under the TRUE reward — so labels contain realistic noise
 * (the better response usually wins, not always).
 *
 * Returns { Xc, Xr } (chosen / rejected), each n×D.
 */
function makePairs(
  n: number,
  uniform: () => number,
  gauss: () => number,
): { Xc: number[][]; Xr: number[][] } {
  const Xc: number[][] = [];
  const Xr: number[][] = [];
  for (let i = 0; i < n; i++) {
    const xA = Array.from({ length: D }, () => gauss());
    const xB = Array.from({ length: D }, () => gauss());
    const pA = 1 / (1 + Math.exp(-(dot(xA, W_TRUE) - dot(xB, W_TRUE))));
    if (uniform() < pA) {
      Xc.push(xA);
      Xr.push(xB);
    } else {
      Xc.push(xB);
      Xr.push(xA);
    }
  }
  return { Xc, Xr };
}

// ---------------------------------------------------------------------------
// Core functions — YOU implement these three (plain arrays only)
// ---------------------------------------------------------------------------

/**
 * Bradley–Terry probability that "chosen" beats "rejected", element-wise:
 *
 *     P(chosen ≻ rejected) = σ(r_chosen − r_rejected)
 *
 * Both inputs are length-n arrays of scalar rewards; return a length-n array.
 * Clamp the exponent argument to [-500, 500] to avoid overflow.
 *
 * TODO: implement.
 *   - Form each reward gap, clamp it, and push it through the sigmoid
 *     σ(z) = 1 / (1 + e^{-z}).
 */
function btProb(rChosen: number[], rRejected: number[]): number[] {
  // TODO: sigmoid of each clamped reward gap
  throw new Error("TODO: implement btProb()");
}

/**
 * The reward-modeling loss (negative log-likelihood of the preferences):
 *
 *     L = −mean log σ(r_chosen − r_rejected)
 *
 * Clip each probability to [1e-12, 1 − 1e-12] before the log.
 *
 * TODO: implement.
 *   - Get the pairwise probabilities from btProb, clip each, and return the
 *     mean of the negative logs.
 */
function btLoss(rChosen: number[], rRejected: number[]): number {
  // TODO: mean negative log of the clipped btProb values
  throw new Error("TODO: implement btLoss()");
}

/**
 * One full-batch gradient-descent step on the Bradley–Terry loss for the
 * linear reward model r_θ(x) = θ·x.
 *
 *   Δ_i    = θ·xChosen_i − θ·xRejected_i               (length n)
 *   ∂L/∂θ  = −(1/n) Σ_i (1 − σ(Δ_i)) · (xChosen_i − xRejected_i)
 *   update:  θ_new = θ − lr · ∂L/∂θ
 *
 * Return the UPDATED θ as a NEW length-D array (do not mutate the input —
 * the training loop reassigns theta = btGradStep(...)).
 *
 * TODO: implement.
 *   - Compute both reward arrays with dot products, then σ(Δ) via btProb.
 *   - Accumulate the gradient: for each pair, add its feature difference
 *     (xChosen − xRejected) scaled by its (1 − σ(Δ_i)) factor; divide by n
 *     and apply the leading minus sign.
 *   - Return the stepped weights.
 */
function btGradStep(
  theta: number[],
  Xc: number[][],
  Xr: number[][],
  lr: number,
): number[] {
  // TODO: build the Bradley–Terry gradient and take one descent step
  throw new Error("TODO: implement btGradStep()");
}

// ---------------------------------------------------------------------------
// Evaluation helpers  (provided — use your bt* functions)
// ---------------------------------------------------------------------------

/** Fraction of held-out pairs where the model ranks chosen above rejected. */
function rankingAccuracy(theta: number[], Xc: number[][], Xr: number[][]): number {
  let correct = 0;
  for (let i = 0; i < Xc.length; i++) {
    if (dot(Xc[i], theta) > dot(Xr[i], theta)) correct++;
  }
  return correct / Xc.length;
}

/** Cosine similarity between two vectors. */
function cosine(a: number[], b: number[]): number {
  return dot(a, b) / (norm(a) * norm(b) + 1e-12);
}

// ---------------------------------------------------------------------------
// Harness  (provided — do not edit)
// ---------------------------------------------------------------------------

function main(): void {
  console.log("Task 2 — Reward model from preferences (Bradley–Terry)\n");

  const uniform = makeRng(SEED);
  const gauss = makeGaussianFrom(makeRng(SEED + 99));
  const { Xc: XcTrain, Xr: XrTrain } = makePairs(N_TRAIN, uniform, gauss);
  const { Xc: XcTest, Xr: XrTest } = makePairs(N_TEST, uniform, gauss);
  console.log(`  Pairs: ${N_TRAIN} train / ${N_TEST} held-out, D=${D}`);
  console.log(`  Hidden true reward w* = [${W_TRUE}]\n`);

  console.log("[1/2] Training the linear reward model on preference pairs...");
  let theta = new Array(D).fill(0) as number[];
  const lossHistory: number[] = [];
  for (let epoch = 0; epoch < EPOCHS; epoch++) {
    lossHistory.push(
      btLoss(
        XcTrain.map((x) => dot(x, theta)),
        XrTrain.map((x) => dot(x, theta)),
      ),
    );
    theta = btGradStep(theta, XcTrain, XrTrain, LR);
  }

  for (const e of [1, 5, 20, 60, EPOCHS]) {
    console.log(
      `  epoch ${String(e).padStart(4)}: BT loss = ${lossHistory[e - 1].toFixed(4)}`,
    );
  }

  const first30 = lossHistory.slice(0, 30);
  let monotonic = true;
  for (let i = 0; i < first30.length - 1; i++) {
    if (first30[i + 1] > first30[i] + 1e-9) monotonic = false;
  }

  console.log("\n[2/2] Evaluating the learned reward model...");
  const acc = rankingAccuracy(theta, XcTest, XrTest);
  const cos = cosine(theta, W_TRUE);
  console.log(`  learned θ = [${theta.map((t) => t.toFixed(3)).join(", ")}]`);
  console.log(`  held-out ranking accuracy = ${acc.toFixed(3)}`);
  console.log(`  cosine(θ, w*)             = ${cos.toFixed(4)}`);
  console.log(`  loss monotone (first 30)  = ${monotonic}`);

  // ── Acceptance checks ────────────────────────────────────────────────────
  console.log("\nAcceptance:");
  const okAcc = acc >= 0.9;
  const okCos = cos >= 0.9;
  const okMonotone = monotonic;
  console.log(
    `  [${okAcc ? "x" : " "}] held-out ranking accuracy >= 0.9  (acc = ${acc.toFixed(3)})`,
  );
  console.log(
    `  [${okCos ? "x" : " "}] cosine(θ, w*) >= 0.9  (cos = ${cos.toFixed(4)})`,
  );
  console.log(
    `  [${okMonotone ? "x" : " "}] BT loss decreases monotonically over first 30 epochs`,
  );

  if (okAcc && okCos && okMonotone) {
    console.log("\n  All acceptance checks passed.");
  } else {
    console.log("\n  Some checks failed — revisit your implementation.");
  }
}

main();

// This file has no imports; the empty export marks it as an ES module so its
// top-level names stay file-scoped (keeps `pnpm typecheck` happy across modules).
export {};

File: modules/13b-alignment/py/02_reward_model.py
"""
Task 2 🟡 — Reward model from preferences (Bradley–Terry).

What you'll learn:
  - The Bradley–Terry model: turning "chosen ≻ rejected" pairs into a
    probabilistic training signal
  - The reward-modeling loss used by InstructGPT: −log σ(r_chosen − r_rejected)
  - Why maximum likelihood on noisy preferences recovers the hidden reward
    (direction) — the exact recipe behind every RLHF reward model

The math (README derives each step):

  Responses are feature vectors x ∈ ℝ^D. A hidden true reward scores them:
      r*(x) = w* · x
  Preference data: for a pair (x_a, x_b), the label "a chosen" is sampled with
      P(a ≻ b) = σ(r*(x_a) − r*(x_b))          (Bradley–Terry)
  We fit a linear reward model  r_θ(x) = θ · x  by minimising, over pairs,
      L(θ) = −mean log σ(Δ) ,   Δ = r_θ(x_chosen) − r_θ(x_rejected)
  Gradient (derive it: d/dΔ [−log σ(Δ)] = −(1 − σ(Δ)), then chain to θ):
      ∂L/∂θ = −mean (1 − σ(Δ)) · (x_chosen − x_rejected)

You implement the three core functions (bt_prob, bt_loss, bt_grad_step).
Data generation, the train/held-out split, the training loop, and the
evaluation report are provided and runnable.

How to run:
  uv run python modules/13b-alignment/py/02_reward_model.py
"""

from __future__ import annotations

import numpy as np

SEED = 21
D = 6  # response feature dimension
N_TRAIN = 400  # preference pairs for training
N_TEST = 200  # held-out preference pairs
LR = 0.2
EPOCHS = 200

# The hidden "true" reward direction the annotators (noisily) follow.
W_TRUE = np.array([3.0, -4.0, 2.0, 1.2, -2.4, 1.6])


# ---------------------------------------------------------------------------
# Preference-pair generation  (provided — do not edit)
# ---------------------------------------------------------------------------


def make_pairs(n: int, rng: np.random.Generator) -> tuple[np.ndarray, np.ndarray]:
    """
    Generate n preference pairs. Each pair draws two random "responses"
    x_a, x_b ~ N(0, I_D); the winner is sampled from the Bradley–Terry
    probability under the TRUE reward — so labels contain realistic noise
    (the better response usually wins, not always).

    Returns (X_chosen, X_rejected), each of shape (n, D).
    """
    X_a = rng.normal(0.0, 1.0, size=(n, D))
    X_b = rng.normal(0.0, 1.0, size=(n, D))
    p_a = 1.0 / (1.0 + np.exp(-(X_a @ W_TRUE - X_b @ W_TRUE)))
    a_wins = rng.random(n) < p_a
    X_chosen = np.where(a_wins[:, None], X_a, X_b)
    X_rejected = np.where(a_wins[:, None], X_b, X_a)
    return X_chosen, X_rejected


# ---------------------------------------------------------------------------
# Core functions — YOU implement these three
# ---------------------------------------------------------------------------


def bt_prob(r_chosen: np.ndarray, r_rejected: np.ndarray) -> np.ndarray:
    """
    Bradley–Terry probability that "chosen" beats "rejected", element-wise:

        P(chosen ≻ rejected) = σ(r_chosen − r_rejected)

    Both inputs are (n,) arrays of scalar rewards; return an (n,) array.
    Clamp the exponent argument to [-500, 500] to avoid overflow.

    TODO: implement.
      - Form the reward gap, clamp it with np.clip, and push it through the
        sigmoid  σ(z) = 1 / (1 + e^{-z}).
    """
    # TODO: sigmoid of the clipped reward gap
    raise NotImplementedError("TODO: implement bt_prob()")


def bt_loss(r_chosen: np.ndarray, r_rejected: np.ndarray) -> float:
    """
    The reward-modeling loss (negative log-likelihood of the preferences):

        L = −mean log σ(r_chosen − r_rejected)

    Clip the probability to [1e-12, 1 − 1e-12] before the log.

    TODO: implement.
      - Get the pairwise probabilities from bt_prob, clip them with np.clip,
        and return the mean negative log as a float.
    """
    # TODO: mean negative log of the clipped bt_prob values
    raise NotImplementedError("TODO: implement bt_loss()")


def bt_grad_step(
    theta: np.ndarray, X_chosen: np.ndarray, X_rejected: np.ndarray, lr: float
) -> np.ndarray:
    """
    One full-batch gradient-descent step on the Bradley–Terry loss for the
    linear reward model r_θ(x) = θ·x.

      Δ_i    = θ·x_chosen_i − θ·x_rejected_i                (shape (n,))
      ∂L/∂θ  = −(1/n) Σ_i (1 − σ(Δ_i)) · (x_chosen_i − x_rejected_i)
      update:  θ_new = θ − lr · ∂L/∂θ

    Return the UPDATED θ as a new (D,) array (do not mutate the input — the
    training loop reassigns theta = bt_grad_step(...)).

    TODO: implement.
      - Compute both reward arrays with matrix-vector products, then σ(Δ) via
        bt_prob.
      - Weight each pair's feature difference (x_chosen − x_rejected) by its
        (1 − σ(Δ_i)) factor — broadcasting a (n, 1) column over (n, D) works —
        and average over pairs with the leading minus sign.
      - Return the stepped weights.
    """
    # TODO: build the Bradley–Terry gradient and take one descent step
    raise NotImplementedError("TODO: implement bt_grad_step()")


# ---------------------------------------------------------------------------
# Evaluation helpers  (provided — use your bt_* functions)
# ---------------------------------------------------------------------------


def ranking_accuracy(theta: np.ndarray, Xc: np.ndarray, Xr: np.ndarray) -> float:
    """Fraction of held-out pairs where the model ranks chosen above rejected."""
    return float(np.mean(Xc @ theta > Xr @ theta))


def cosine(a: np.ndarray, b: np.ndarray) -> float:
    """Cosine similarity between two vectors."""
    return float(a @ b / (np.linalg.norm(a) * np.linalg.norm(b) + 1e-12))


# ---------------------------------------------------------------------------
# Harness  (provided — do not edit)
# ---------------------------------------------------------------------------


def main() -> None:
    print("Task 2 — Reward model from preferences (Bradley–Terry)\n")

    rng = np.random.default_rng(SEED)
    Xc_train, Xr_train = make_pairs(N_TRAIN, rng)
    Xc_test, Xr_test = make_pairs(N_TEST, rng)
    print(f"  Pairs: {N_TRAIN} train / {N_TEST} held-out, D={D}")
    print(f"  Hidden true reward w* = {W_TRUE}\n")

    print("[1/2] Training the linear reward model on preference pairs...")
    theta = np.zeros(D)
    loss_history: list[float] = []
    for _epoch in range(EPOCHS):
        loss_history.append(bt_loss(Xc_train @ theta, Xr_train @ theta))
        theta = bt_grad_step(theta, Xc_train, Xr_train, LR)

    for e in (1, 5, 20, 60, EPOCHS):
        print(f"  epoch {e:>4}: BT loss = {loss_history[e - 1]:.4f}")

    first30 = loss_history[:30]
    monotonic = all(first30[i + 1] <= first30[i] + 1e-9 for i in range(len(first30) - 1))

    print("\n[2/2] Evaluating the learned reward model...")
    acc = ranking_accuracy(theta, Xc_test, Xr_test)
    cos = cosine(theta, W_TRUE)
    print(f"  learned θ = {np.round(theta, 3)}")
    print(f"  held-out ranking accuracy = {acc:.3f}")
    print(f"  cosine(θ, w*)             = {cos:.4f}")
    print(f"  loss monotone (first 30)  = {monotonic}")

    # ── Acceptance checks ────────────────────────────────────────────────────
    print("\nAcceptance:")
    ok_acc = acc >= 0.9
    ok_cos = cos >= 0.9
    ok_monotone = monotonic
    print(f"  [{'x' if ok_acc else ' '}] held-out ranking accuracy >= 0.9  (acc = {acc:.3f})")
    print(f"  [{'x' if ok_cos else ' '}] cosine(θ, w*) >= 0.9  (cos = {cos:.4f})")
    print(f"  [{'x' if ok_monotone else ' '}] BT loss decreases monotonically over first 30 epochs")

    if ok_acc and ok_cos and ok_monotone:
        print("\n  All acceptance checks passed.")
    else:
        print("\n  Some checks failed — revisit your implementation.")


if __name__ == "__main__":
    main()
